rfcs/0128-redeployable-clients.md

```diff
@@ -0,0 +1,393 @@
+taskcluster deployment team needing to restart or rebuild their taskcluster
+deployment. The second team can make their API references available and modify
+them as they see fit while the taskcluster core platform is running without any
+maintenance overhead.
+
+But we may wish to implement such a thing in a future PR - for the time being,
+the author does not care about the publish mechanism, only that the content is
+eventually served.
+
+## Changes to taskcluster client building procedure
+
+### All language clients (go, java, node.js, python)
+
+The build process for these clients should be moved into a standalone tool that
+requires a `TASKCLUSTER_ROOT_URL` to be specified to build the client for. It
+will then query the API References manifest in order to determine all the APIs
+it needs to build, and from there build packages for each API reference it
+finds.
+
+Please note, the built client will retain NO REFERENCES to the client it was
+built from, other than as comments as a reference. The value of the
+`TASKCLUSTER_ROOT_URL` at build time is simply to freeze the API references and
+schemas to a given set. A client can be built against one taskcluster
+deployment, but calling code may choose to use that client to talk to a
+different cluster. It is the responsibility of the caller to ensure that the
+methods he/she calls are compatible with the deployment he/she connects to.
+
+The API reference for a service must declare the `$schema` json schema property
+to say which API reference it implements. If this matches an API reference that
+the client builder knows how to build, it will build a source code package for
+it in a local directory. If not, it may either skip it with a warning message,
+or throw an error.
+
+This code for the code generator will live in a taskcluster repo, and only
+needs to be updated when the code generation process changes, not when service
+API definitions change.
+
+### Language clients wihtout type code generation (node.js, python)
+
+The clients that do not dynamically generate code still need to retain
+information from the API references and schemas. Those clients should be
+shipped together with a set of references and schemas, but provide the facility
+for these references and schemas to be refetched from a taskcluster deployment
+(specified by a taskcluster root URL).
+
+## Changes to taskcluster client features and configuration
+
+Constructors for clients should require the rootURL to be explicitly provided,
+i.e. no default. This is in order to ensure that the choice of deployed cluster
+to connect to is an active one. The client may provide a utility methods to fetch
+TASKCLUSTER_ROOT_URL from the environment, but it should not be assumed that if
+this environment variable exists, that it should be used. Rather, code that calls
+the taskcluster client should have to explicitly declare they wish to do so, for
+example:
+
+```
+queue := queue.NewFromEnvVars()
```

Comment:
I guess for node.js and python, that wouldn't be the end of the world. The problem comes when you have static language clients where the client gets compiled into some other binary that get used in other places (for example, the taskcluster-client is used internally by generic-worker, but also by taskcluster-proxy, for example). In the case of generic worker, for example, it will fetch the root URL from the generic-worker.config file, rather than the TASKCLUSTER_ROOT_URL environment variable. If it were for some reason set in the environment, and then it took it from there, it might not necessarily be the correct TASKCLUSTER_ROOT_URL to use.

Also consider the situation where e.g. somebody writes a tool that syncs role definitions between two taskcluster deployments. If internally the client uses the environment variable TASKCLUSTER_ROOT_URL to discover which cluster it needs to talk to, only one can be specified in that process. It kind of assumes it is only talking to one deployment.

In a command line tool where you know what your tool is and what it is being used for, and have a feeling about the context it might be called from, it seems reasonable to take config from environment variables (e.g. for writing a web service). But when it isn't a service or command line application you are writing, but an internal library, it feels cleaner and more appropriate to take configuration as parameters passed to methods and functions. By providing a utility method to fetch config from env vars, you expose the possibility to the calling code to configure based on env vars, but you don't _force_ it. Then it is up to the author of the command line tool / service / etc that calls your library to decide explicitly whether they want that or not. Then another advantage of being explicit is it is a little less magical and easier to reason about when you are degugging that code and trying to work out where it got the configuration from. I think. :-)

Reply:
I think we all agree here -- libraries shouldn't be reading env variables without explicit instruction from the caller.

We just need to roll out that change gently since it will break every user of the client in possibly-subtle ways.

Reply:
lol, I guess I misread @imbstack's comment, and thought he meant we should support only accepting env vars, but now I see he meant the opposite, so I didn't need to write all that! 😆

Reply:
I *thought* we all agreed that *all* env vars will no longer be automatically read by the clients, but the text of the RFC still refers only to `TASKCLUSTER_ROOT_URL`.  I think clients should be consistent in which env vars they read, and they currently read `TASKCLUSTER_CLIENT_ID` etc.

This will be a big breaking change for our users (and I've held @tomprince back from using newer clients in anticipation), so I think we should be explicit about it and make sure they get a chance to comment here.  Alternately, we could remove this bit from this RFC and move it to a new one.